site/docs/expressions/scalar_functions.md

```diff
@@ -43,10 +43,67 @@ For scalar functions (one value produced for each record), a function binding de
 
 
 
+## Argument Types & Output Type Derivation
+
+### Input Argument Types
+
+Input arguments can be declared in one of two ways: materialized or parameterized.
+
+* Materialized Type: Materialized types are either simple types or compound types with all parameters fully defined (without any parameters). Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc.
+* Parameterized types: Types can be parameterized such that the parameter can be used in a output derivation e.g. `f(K) => K` or to guarantee consistency of type between input arguments `f(K,K) => boolean`. Example parameterized types would be VARCHAR(T), List&lt;E&gt;, MAP&lt;K, fp32&gt;, etc. A parameters is named with a simple UTF8 character or string.
+
+### Direct Return Types
+
+A direct return type is one that is fully known at function definition type. Example simple direct return types would be things such as i32, fp32. For compound types, a direct return type must be fully declared. Example fully defined types: VARCHAR(20), DECIMAL(25,5)
+
+### Return Type Expressions
+
+Any function can declare a return type expression. A return type expression uses a simplified set of expressions to describe how the return type should be returned. For example, a return expression could be as simple as the return of parameter declared in the arguments. For example `f(List<K>) =>K ` or can be a simple mathematical or conditional expression such as `add(decimal(a,b), decimal(c,d)) => decimal(a+c, b+d)`. For the simple expression language, there is a very narrow set of types:
+
+* Integer: 64 bit signed integer (can be a literal or a parameter value)
+* Boolean: True and False
+* Type: A Substrait type (with possibly additional embedded expressions)
+
+These types are evaluated using a small set of operations to support common scenarios. Since it is possible that output derivation ultimately determines that binding is not allowed, a `not_bindable` operation is provided that states that the particular binding is not allowed. List of valid operations:
+
+```
+Math: +, -, *, /, min, max
+Boolean: &&, ||, !, <, >, ==
+Input Parameters: type parameter, literal parameter
+Literals: integer, type
+Special: !bindable
+```
+
+Fully defined with argument types:
+
+* `not(boolean x) => boolean` 
+* `and(boolean a, boolean b) => boolean` 
+* `or(boolean a, boolean b) => boolean` 
+* `multiply(integer a, integer b) => integer`
+* `divide(integer a, integer b) => integer`
+* `add(integer a, integer b) => integer`
+* `substract(integer a, integer b) => integer`
+* `min(integer a, integer b) => integer`
+* `max(integer a, integer b) => integer`
+* `equal(integer a, integer b) => boolean`
+* `greater_than(integer a, integer b) => boolean`
+* `less_than(integer a, integer b) => boolean`
+* `equal(Type a, Type b) => boolean`
+* `if(boolean a) then (integer) else (integer)`
+* `if(boolean a) then (type) else (type)`
+* `not_bindable()`
```

Comment:
It looks like there's `!bindable` and `not_bindable()`. Are these the same thing?

Reply:
Oh, is `not_bindable()` what is produced by the syntax `!bindable`?

Reply:
None of this is syntax, just concepts. Syntax would be delegated to the serialization formats. In the abbreviated section I try to make it easy to understand what the ops are. The section below is blown out to better describe the parameters, etc.